keras_cv/training/contrastive/contrastive_trainer.py

```diff
@@ -0,0 +1,159 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+from keras_cv.layers import preprocessing
+
+
+class ContrastiveTrainer(keras.Model):
+    """Creates a self-supervised contrastive trainer for a model.
+
+    Args:
+        encoder: a `keras.Model` to be pre-trained. In most cases, this encoder
+            should not include a top dense layer.
+        augmenter: a preprocessing layer to randomly augment input images for contrastive learning.
+        projector: a projection model for contrastive training
+        include_probe: Whether to include a single fully-connected layer during
+            training for probing classification accuracy using the learned encoding.
+            Note that this should be specified iff training with labeled images.
+            If provided, `classes` must be provided.
+        probe_metrics: a list of metrics for the linear probe, only used if `include_probe` is True.
+        classes: optional number of classes to classify images into, only to be
+            specified if `include_probe` is True.
+
+    Returns:
+      A `keras.Model` instance.
+
+    """
+
+    def __init__(
+        self,
+        encoder,
+        augmenter,
+        projector,
+        include_probe,
+        probe_metrics=[keras.metrics.CategoricalAccuracy(name="probe_accuracy")],
+        classes=None,
+    ):
+        super().__init__()
+
+        if encoder.output.shape.rank != 2:
+            raise ValueError("Encoder must have a flattened output")
+
+        if include_probe:
+            if not classes:
+                raise ValueError(
+                    "`classes` must be specified when `include_probe` is `True`."
+                )
+
+        self.include_probe = include_probe
+
+        self.augmenter = augmenter
+        self.encoder = encoder
+        self.projector = projector
+
+        self.loss_metric = keras.metrics.Mean(name="loss")
+        self.probe_loss_metric = keras.metrics.Mean(name="probe_loss")
+        self.probe_metrics = probe_metrics or []
+
+        if self.include_probe:
+            self.probing_top = layers.Dense(classes, name="linear_probe")
+
+    def compile(self, optimizer, loss, probe_optimizer=None, **kwargs):
+        super().compile(**kwargs)
+
+        if self.include_probe and not probe_optimizer:
+            raise ValueError(
+                "`probe_optimizer` must be specified when `include_probe` is `True`."
+            )
+
+        self.optimizer = optimizer
+        self.loss = loss
+
+        if self.include_probe:
+            self.probe_loss = keras.losses.CategoricalCrossentropy(from_logits=True)
```

Comment:
should probe_loss just be a keyword argument to compile()?

Reply:
Yeah I think that's a better plan -- done.